src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java

```diff
@@ -841,80 +956,182 @@ else if (block instanceof BlockEndPortal || block instanceof BlockMobSpawner ||
+            {
+                final int slot = worker.findFirstSlotInInventoryWith(tempStack.getItem(), tempStack.getItemDamage());
+                if (slot != -1)
+                {
+                    new InvWrapper(getInventory()).extractItem(slot, 1, false);
+                    reduceNeededResources(tempStack);
+                }
+            }
+        }
+
+        if (Configurations.builderBuildBlockDelay > 0 && block != Blocks.AIR)
         {
-            getInventory().decrStackSize(slot, 1);
-            reduceNeededResources(block);
+            setDelay(Configurations.builderBuildBlockDelay);
         }
+
         return true;
     }
 
     /**
      * Reduces the needed resources by 1.
      *
-     * @param block the block which has been used now.
+     * @param stack the stack which has been used now.
      */
-    public void reduceNeededResources(final Block block)
+    public void reduceNeededResources(final ItemStack stack)
     {
         final AbstractBuilding workerBuilding = this.getOwnBuilding();
         if (workerBuilding instanceof BuildingBuilder)
         {
-            ((BuildingBuilder) workerBuilding).reduceNeededResource(block, 1);
+            ((BuildingBuilder) workerBuilding).reduceNeededResource(stack, 1);
         }
     }
 
-    private void spawnEntity(@Nullable final Entity entity)
+    /**
+     * Get the entity of an entityInfo object.
+     *
+     * @param entityInfo the input.
+     * @return the output object or null.
+     */
+    @Nullable
+    private Entity getEntityFromEntityInfoOrNull(Template.EntityInfo entityInfo)
     {
-        if (entity != null && job instanceof JobBuilder)
+        try
         {
-            final BlockPos pos = ((JobBuilder) job).getStructure().getOffsetPosition();
-
-            if (entity instanceof EntityHanging)
-            {
-                @NotNull final EntityHanging entityHanging = (EntityHanging) entity;
+            return EntityList.createEntityFromNBT(entityInfo.entityData, world);
+        }
+        catch (RuntimeException e)
+        {
+            Log.getLogger().info("Couldn't restore entitiy", e);
+            return null;
+        }
+    }
 
-                entityHanging.posX += pos.getX();
-                entityHanging.posY += pos.getY();
-                entityHanging.posZ += pos.getZ();
-                //also sets position based on tile
-                entityHanging.setPosition(
-                        entityHanging.getHangingPosition().getX(),
-                        entityHanging.getHangingPosition().getY(),
-                        entityHanging.getHangingPosition().getZ());
+    /**
+     * Set the currentStructure to null.
+     */
+    public void resetCurrentStructure()
+    {
+        currentStructure = null;
+    }
 
-                entityHanging.setWorld(world);
-                entityHanging.dimension = world.provider.getDimension();
+    private Boolean spawnEntity(@NotNull final Structure.StructureBlock currentBlock)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') The Cyclomatic Complexity of this method "spawnEntity" is 16 which is greater than 15 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 43 to the 20 allowed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)